Correct and usable normal distribution sampler

bugQ opened the pull request:
`normal` and `standardNormal` were severely broken based on sampling.
Not sure what algorithm was used, but this one is Box-Muller, which is very simple and proven.
The only disadvantage is that `normal` does not have bounded output, but this is acceptable for most uses of normally distributed variables, and has a more typical interface.

bugQ committed:
Correct and usable normal distribution sampler

`normal` and `standardNormal` were severely broken based on sampling.
Not sure what algorithm was used, but this one is Box-Muller, which is very simple and proven.
The only disadvantage is that `normal` does not have bounded output, but this is acceptable for most uses of normally distributed variables, and has a more typical interface.

bugQ committed:
missing )

bugQ committed:
[style] remove redundancy

mgold commented:
Thanks for the PR! This is kinda over my head but the code looks simple enough. I'll have time to investigate further over the weekend. It will also require a major version bump, so that implies a bit of scrutiny, planning, and coordination.

bugQ commented:
Here's some info on the [Box-Muller transform](https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform).  It's pretty common way of generating numbers on the standard normal distribution from uniforms, and might actually be the most efficient in this case.  Other options usually require rejection sampling, i.e. repeated sampling of uniforms until they meet certain conditions, which seems tricky to do with Elm's generators.

mgold commented:
[Tell me about it...](https://github.com/mgold/elm-random-pcg/blob/49eebd49724884861267efbe23f8631f3118aac1/src/Random/Pcg.elm#L232-L247)

Okay, so this looks good except the `(1 - u)` term. Why not just use `u`? Both expressions will be uniformly distributed on [0,1].

bugQ commented:
That was based on the assumption that the range of `Random.float 0 1` doesn't actually include 1, which is the case for most RNG implementations, because of the asymmetry of 2's complement.  Thus, `(1 - u)` is meant to turn the [0,1) range into (0,1], avoiding ln(0) = NaN.

Looking at [core](https://github.com/elm-lang/core/blob/4.0.4/src/Random.elm#L144-L160), that assumption was correct, but it also looks like it could generate a number slightly below the minimum (about -1.16e10)!  If that's not intended behavior for `Random.float`, then my `standardNormal` would possibly need to be changed along with it.

bugQ commented:
Ah yes, in fact, that might cause problems with this implementation regardless, because if `u` is even slightly negative, then `sqrt (logBase e (1 - u))` is not a real number.  Okay, maybe this should be postponed.

mgold commented:
Oh no! Here's what I recommend:
- Write the best normal generator you can with the current implementation of `float` core. If that means throwing a `floor` into this PR, do it.
- Attempt to reproduce the issue using [elm-random-pcg](https://github.com/mgold/elm-random-pcg), of which I am also the author. If you can, open an issue or PR and we can fix it. If it doesn't repro, it's probably not worth fixing in core, since that library will be replacing core's implementation at some point.

bugQ commented:
Oh, that's good to know.  That indeed looks like a much better `float` implementation.  Substituting `minInt` and `maxInt` for the peeled values produces 0 and 0.9999999999999999, as expected.

I suppose I was a bit too hasty in opening an issue on core.  I'll amend this PR as well as that issue.

mgold commented:
I wrote a program to find the bounds of what `float 0 1` produces over ten million numbers, and core produces ( -1.1641532182693481e-10, 0.9999997616978362) and PCG produces (3.2658150628250837e-9, 0.9999999993300184). So I have both reproduced the issue in core and failed to reproduce it in PCG. That said, I don't think you did the wrong thing by opening the issue!

So, it seems that in the mean time the `(1 - u)` trick will help avert disaster if core produces a slightly negative number, since it will become a number slightly greater than one. So does that mean this is good to merge?

bugQ committed:
fix standardNormal

bugQ committed:
patch standardNormal until Random.float is fixed

Random.float can currently produce a value slightly below the given range, which would result in NaN here.

bugQ commented:
The above fixes were still necessary because there was an error with the parentheses and because a number slightly greater than one would produce a NaN from the square root.  But, now it should be good to go.

bugQ commented:
... I am so sorry.  There's still a missing ).

mgold commented:
Got it, no worries.
